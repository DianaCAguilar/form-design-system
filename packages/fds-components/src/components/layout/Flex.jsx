import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import childrenOfType from '../../util/propValidation';

/**
 * @description map of alignment prop values to class names
 */
const DirectionPropMap = {
  row: 'flex--row',
  column: 'flex--column',
};

/**
 * @description map of justify prop values to class names
 */
const JustifyPropMap = {
  flexEnd: 'flex--justifyEnd',
  center: 'flex--justifyCenter',
  spaceBetween: 'flex--justifySpaceBetween',
  spaceAround: 'flex--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
const AlignPropMap = {
  flexStart: 'flex--alignStart',
  center: 'flex--alignCenter',
  flexEnd: 'flex--alignEnd',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Flex = (props) => {
  const classNames = cx(
    'flex',
    DirectionPropMap[props.direction],
    {
      [`flex--${props.direction}--reverse`]: props.reverse,
    },
    {
      [JustifyPropMap[props.justify]]: props.justify,
      [AlignPropMap[props.align]]: props.align,
    },
    {
      'flex--wrap': props.wrap,
      'flex--noGutters': props.noGutters,
    },
    props.className
  );

  return <div className={classNames}>{props.children}</div>;
};

Flex.defaultProps = {
  direction: 'row',
  noGutters: false,
};

Flex.propTypes = {
  direction: PropTypes.oneOf(Object.keys(DirectionPropMap)),
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),
  noGutters: PropTypes.bool,
  wrap: PropTypes.bool,
  reverse: PropTypes.bool,
  className: PropTypes.string,
  children: childrenOfType(['FlexItem']),
};

export default Flex;
