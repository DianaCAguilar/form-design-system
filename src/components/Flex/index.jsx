import React from 'react';
import cc from 'classcat';
import PropTypes from 'prop-types';

/**
 * @description map of alignment prop values to class names
 */
export const DirectionPropMap = {
  row: 'flex--row',
  column: 'flex--column',
};

/**
 * @description valid breakpoint values
 */
export const SwitchDirectionProps = ['xs', 's', 'm', 'l', 'xl'];

/**
 * @description map of justify prop values to class names
 */
export const JustifyPropMap = {
  end: 'flex--justifyEnd',
  center: 'flex--justifyCenter',
  spaceBetween: 'flex--justifySpaceBetween',
  spaceAround: 'flex--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
export const AlignPropMap = {
  start: 'flex--alignStart',
  end: 'flex--alignEnd',
  center: 'flex--alignCenter',
  stretch: 'flex--alignStretch',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Flex = ({
  align = 'stretch',
  direction = 'row',
  switchDirection: breakpoint,
  reverse,
  justify,
  noGutters,
  children,
}) => {
  const classNames = cc([
    'flex',
    AlignPropMap[align],
    DirectionPropMap[direction],
    {
      [`${breakpoint}:flex--${direction === 'row' ? 'column' : 'row'}`]: breakpoint,
    },
    {
      'flex--reverse': reverse,
    },
    {
      [JustifyPropMap[justify]]: justify,
    },
    {
      'flex--noGutters': noGutters,
    },
  ]);

  return <div className={classNames}>{children}</div>;
};

Flex.propTypes = {
  /** sets flex-direction (along with either 100% height or width) */
  direction: PropTypes.oneOf(Object.keys(DirectionPropMap)),

  /** switches flex-direction at a given breakpoint */
  switchDirection: PropTypes.oneOf(SwitchDirectionProps),

  /** sets standard justify-content */
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),

  /** sets standard align-items */
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),

  /** When set, gutters are removed from `FlexItem` children */
  noGutters: PropTypes.bool,

  /** When set, `FlexItem` order is rendered in reverse */
  reverse: PropTypes.bool,

  /** React children (should be of type `FlexItem`) */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
    .isRequired,
};

export default Flex;
