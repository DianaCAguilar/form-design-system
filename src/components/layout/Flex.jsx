import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

/**
 * @description map of alignment prop values to class names
 */
export const DirectionPropMap = {
  row: 'flex--row',
  column: 'flex--column',
};

/**
 * @description map of justify prop values to class names
 */
export const JustifyPropMap = {
  end: 'flex--justifyEnd',
  center: 'flex--justifyCenter',
  spaceBetween: 'flex--justifySpaceBetween',
  spaceAround: 'flex--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
export const AlignPropMap = {
  start: 'flex--alignStart',
  end: 'flex--alignEnd',
  center: 'flex--alignCenter',
  stretch: 'flex--alignStretch',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Flex = (props) => {
  const classNames = cx(
    'flex',
    AlignPropMap[props.align],
    DirectionPropMap[props.direction],
    {
      [`flex--${props.direction}--reverse`]: props.reverse,
    },
    {
      [JustifyPropMap[props.justify]]: props.justify,
    },
    {
      'flex--wrap': props.wrap,
      'flex--noGutters': props.noGutters,
    }
  );

  return <div className={classNames}>{props.children}</div>;
};

Flex.defaultProps = {
  direction: 'row',
  align: 'stretch',
};

Flex.propTypes = {
  /** sets flex-direction (along with either 100% height or width) */
  direction: PropTypes.oneOf(Object.keys(DirectionPropMap)),

  /** sets standard justify-content */
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),

  /** sets standard align-items */
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),

  /** When set, gutters are removed from `FlexItem` children */
  noGutters: PropTypes.bool,

  /** When set, `FlexItem` children are allowed to wrap */
  wrap: PropTypes.bool,

  /** When set, `FlexItem` order is rendered in reverse */
  reverse: PropTypes.bool,

  /** React children (should be of type `FlexItem`) */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
    .isRequired,
};

export default Flex;
