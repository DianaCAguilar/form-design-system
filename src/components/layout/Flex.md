# Flex

Flex is an abstraction of CSS flexbox that provides a subset of flexbox functionality.

**A `Flex` should only contain `FlexItem` children**.

## Flex direction
`Flex` renders in row orientation by default. Use the `direction` prop to change flex direction. To reverse the row or column, use the boolean prop `reverse`.

## Flex alignment/justification
Following the flexbox spec, `align` controls cross-axis alignment and `justify` controls main axis alignment. In this example, both `align` and `justify` are set to `end`. The child
`FlexItem` components here are set to `shrink` to their content width.

_See `FlexItem` documentation for more information on the `FlexItem` `shrink` prop_

