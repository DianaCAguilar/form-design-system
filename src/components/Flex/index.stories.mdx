import React from 'react';
import FlexItem from 'components/FlexItem';
import Flex from '.';

const parentStyle = {
  width: '600px',
};

export const Primary = (args) => (
  <div style={parentStyle} className="debug--flex">
    <Flex {...args}>
      <FlexItem shrink>
        <img src="https://place-hold.it/40x40/a02385/fff&text=IMG" />
      </FlexItem>
      <FlexItem shrink>
        <h4>Look at this cool content</h4>
      </FlexItem>
      <FlexItem shrink>
        <p className="color--link">edit</p>
      </FlexItem>
    </Flex>
  </div>
)

export const classicMediaBlock = (args) => (
  <div style={parentStyle} className="debug--flex">
    <Flex {...args}>
      <FlexItem shrink>
        <img src="https://place-hold.it/40x40/a02385/fff&text=:)" />
      </FlexItem>
      <FlexItem>
        Look at this media block. Notice how this area grows to fill the remaining
        space of the fixed width parent container. When this content wraps, it will
        not wrap below the figure the same way it would when positioning the image
        with a<code>float</code>.
      </FlexItem>
    </Flex>
  </div>
)

classicMediaBlock.parameters = {
  docs: {
    description: {
      story:
        'Making the image `FlexItem` `shrink` and allowing the description `FlexItem` to grow creates the classic "media block" layout pattern.',
    },
  },
};

export const autoSizedContentArea = (args) => (
  <div style={{ height: '300px' }} className="debug--flex">
    <Flex direction="column" {...args}>
      <FlexItem shrink>
        <p>
          This is the header, in a <code>FlexItem</code> set to <code>shrink</code>
        </p>
      </FlexItem>
      <FlexItem>
        <p>
          This is the content area, in a <code>FlexItem</code> that will grow to fill
          remaining space
        </p>
      </FlexItem>
      <FlexItem shrink>
        <p>
          This is the footer, in a <code>FlexItem</code> set to <code>shrink</code>
        </p>
      </FlexItem>
    </Flex>
  </div>
)

autoSizedContentArea.parameters = {
  docs: {
    description: {
      story:
        'Within a fixed height container, this `Flex` creates a fixed height header and footer which allows the content area to expand to fill the remaining space.',
    },
  },
};

export default {
  component: Flex,
  subcomponents: { FlexItem },
  title: 'components/Flex',
  parameters: {
    docs: {
      description: {
        component: 'Flex is an abstraction of CSS flexbox that provides a subset of flexbox functionality. **A `Flex` should only contain `FlexItem` children**.',
      },
    },
  },
};