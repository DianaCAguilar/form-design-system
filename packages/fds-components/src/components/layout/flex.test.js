import React from 'react';
import { shallow } from 'enzyme';

import Flex from './Flex';

describe('Flex component', () => {
  const flex = shallow(<Flex />);

  it('matches snapshot', () => {
    expect(flex).toMatchSnapshot();
  });
});
