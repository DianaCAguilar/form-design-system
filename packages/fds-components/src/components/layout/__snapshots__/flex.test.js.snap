// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex component matches snapshot (default props) 1`] = `
<div
  className="flex flex--alignStretch flex--row"
  data-test=""
>
  <FlexItem
    dataTest=""
    shrink={false}
  />
</div>
`;

exports[`Flex component matches snapshot (set all props) 1`] = `
<div
  className="flex flex--alignEnd flex--column flex--column--reverse flex--justifySpaceBetween flex--wrap flex--noGutters foo"
  data-test=""
>
  <FlexItem
    dataTest=""
    shrink={false}
  />
</div>
`;
